src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,480 @@
+    return E + 2.0 * atan2(beta * sin(E), 1 - beta * cos(E));
+
+    // return 2.0 * atan2(std::sqrt(1 + ecc) * sin(E / 2.0), std::sqrt(1 - ecc) * cos(E / 2.0));
+}
+
+/**
+ * @brief Convert hyperbolic eccentric anomaly (E) to true anomaly (ν)
+ */
+static double hyperbolic_eccentric_anomaly_to_true_anomaly(double const &ecc, double const &H)
+{
+    return 2 * atan(sqrt((ecc + 1) / (ecc - 1)) * tanh(H / 2));
+}
+
+/**
+ * @brief Convert true anomaly (ν) to eccentric anomaly (E) for elliptic orbits
+ */
+static double true_anomaly_to_elliptic_eccentric_anomaly(double const &ecc, double const &v)
+{
+    return 2.0 * atan(std::sqrt((1 - ecc) / (1 + ecc)) * tan(v / 2.0));
+}
+
+/**
+ * @brief Convert true anomaly (ν) to hyperbolic eccentric anomaly (E)
+ */ 
+static double true_anomaly_to_hyperbolic_eccentric_anomaly(double const &ecc, double const &v)
+{
+    return 2.0 * atanh(std::clamp(std::sqrt((ecc - 1) / (ecc + 1)) * tan(v / 2.0), -1.0, 1.0));
+}
+
+/**
+ * @brief Compute mean motion 
+ */
+static double mean_motion(double const mu, double const a)
+{
+    return sqrt(std::abs(mu / (a * a * a)));
+}
+
+/**
+ * @brief Compute the radius of an orbiting body at a given true anomaly
+ */
+static double get_orbiting_radius(double const ecc, double const semiMajorAxis, double const trueAnomaly)
+{
+    return semiMajorAxis * (1 - ecc * ecc) / (1 + ecc * cos(trueAnomaly));
+}
+
+/**
+ * @brief Compute the elliptic eccentric anomaly at a given time since epoch
+ */ 
+static double time_to_elliptic_eccentric_anomaly(double const M0, double const eccentricity, double const semiMajorAxis, double const mu, double const timeSinceEpoch)
+{
+    double const nu = mean_motion(mu, semiMajorAxis);
+    double Mt = M0 + timeSinceEpoch * nu;
+    Mt = wrap_angle(Mt);
+    return solve_kepler_elliptic(Mt, eccentricity, 20000);
+}
+
+/**
+ * @brief Compute the hyperbolic eccentric anomaly at a given time since epoch
+ */
+static double time_to_hyperbolic_eccentric_anomaly(double const M0, double const eccentricity, double const semiMajorAxis, double const mu, double const timeSinceEpoch)
+{
+    double const nu = mean_motion(mu, -semiMajorAxis);
+    double const Mt = M0 + timeSinceEpoch * nu;
+    return solve_kepler_hyperbolic(Mt, eccentricity, 20000);
+}
+
+/**
+ * @brief Compute the position and velocity vectors of an orbiting body at the given anomaly
+ */
+static void get_state_vectors(double   const eccentricAnomaly, 
+                              double   const trueAnomaly, 
+                              double   const eccentricity, 
+                              double   const semiMajorAxis, 
+                              double   const gravitationalParameter, 
+                              Vector3d const u, 
+                              Vector3d const v, 
+                              Vector3d const w,
+                              Vector3d      &position, 
+                              Vector3d      &velocity)
```

Comment:
it doesn't appear that position and velocity are used as input parameters, but are just output parameters.

Consider returning `std::pair<Vector3d, Vector3d>` instead, and then unpacking the tuple at the place where the function is called.

This is less about codegen and more about just how using the function looks.